frame/contracts/src/lib.rs

```diff
@@ -1035,40 +1101,55 @@ impl<T: Trait + Send + Sync> SignedExtension for CheckBlockGasLimit<T> {
 	type Call = <T as Trait>::Call;
 	type AdditionalSigned = ();
 	type DispatchInfo = DispatchInfo;
-	type Pre = ();
+	/// Optional pre-dispatch check data.
+	///
+	/// It is present only for the contract calls that operate with gas.
+	type Pre = Option<DynamicWeightData<T::AccountId, NegativeImbalanceOf<T>>>;
 
-	fn additional_signed(&self) -> sp_std::result::Result<(), TransactionValidityError> { Ok(()) }
+	fn additional_signed(&self) -> Result<(), TransactionValidityError> { Ok(()) }
+
+	fn pre_dispatch(
+		self,
+		who: &Self::AccountId,
+		call: &Self::Call,
+		_: DispatchInfo,
+		_: usize,
+	) -> Result<Self::Pre, TransactionValidityError> {
+		Self::perform_pre_dispatch_checks(who, call)
+			.map_err(Into::into)
+	}
 
 	fn validate(
 		&self,
-		_: &Self::AccountId,
+		who: &Self::AccountId,
 		call: &Self::Call,
 		_: Self::DispatchInfo,
 		_: usize,
 	) -> TransactionValidity {
-		let call = match call.is_sub_type() {
-			Some(call) => call,
-			None => return Ok(ValidTransaction::default()),
-		};
+		Self::perform_pre_dispatch_checks(who, call)
+			.map(|_| ValidTransaction::default())
+	}
 
-		match call {
-			Call::claim_surcharge(_, _) | Call::update_schedule(_) =>
-				Ok(ValidTransaction::default()),
-			Call::put_code(gas_limit, _)
-				| Call::call(_, _, gas_limit, _)
-				| Call::instantiate(_, gas_limit, _, _)
-			=> {
-				// Check if the specified amount of gas is available in the current block.
-				// This cannot underflow since `gas_spent` is never greater than `T::BlockGasLimit`.
-				let gas_available = T::BlockGasLimit::get() - <Module<T>>::gas_spent();
-				if *gas_limit > gas_available {
-					// gas limit reached, revert the transaction and retry again in the future
-					InvalidTransaction::ExhaustsResources.into()
-				} else {
-					Ok(ValidTransaction::default())
-				}
-			},
-			Call::__PhantomItem(_, _)  => unreachable!("Variant is never constructed"),
+	fn post_dispatch(pre: Self::Pre, _info: DispatchInfo, _len: usize) {
+		if let Some(DynamicWeightData {
+			transactor,
+			imbalance,
+		}) = pre {
+			let (gas_left, gas_spent) = GasUsageReport::take();
+
+			// These should be OK since we don't buy more
+			let unused_weight = gas_left as Weight;
+			let spent_weight = gas_spent as Weight;
+
+			let refund = T::WeightToFee::convert(unused_weight);
```

Comment:
I don't think this is right with a non-linear weight model. Say `L` is the weight limit, `S` is the amount spent, and `U` is unused. Then clearly `L = S + U`. But now say you have a quadratic weight function `fee(weight)`. Then you first withdraw `fee(L)` then deposit back `fee(U)`. But really you wanted to just withdraw `fee(S)`. But `fee(S) = fee(L) - fee(U)` only if `fee` is linear.

Instead, the refund should just be `imbalance - T::WeightToFee::convert(spent_weight)`.

Reply:
Good point, for now there is assumption that weight is linear. As you've noticed this is not the only place where the code assumes that.

I think there is more to it than just this and we might need to change the weight mechanism a bit. That said, I think we might just proceed with this assumption, documenting that the contracts module is compatible with linear weight model at the moment)

Thanks for the idea of fix though, might be useful